feat: open topics in new tab; deprecate unused config props; add unload

Blackbaud-ColbyWhite opened the pull request.

Blackbaud-ColbyWhite committed:
feat: move config merge logic from bb-help spa

Blackbaud-ColbyWhite committed:
ci: list new.tab as a build branch

Blackbaud-ColbyWhite committed:
feat: open topics in new tab; remove comms logic

Blackbaud-ColbyWhite committed:
feat: more elaborate local demo for testing

Blackbaud-ColbyWhite committed:
feat: add unload func

Blackbaud-ColbyWhite committed:
feat: add locale selection to demo

Blackbaud-ColbyWhite committed:
npm version prerelease --preid=rc

Blackbaud-ColbyWhite committed:
ci: use skyux linting rules

tslint config taken from
https://github.com/blackbaud/skyux-sdk-builder/blob/71f6bf4af2267b31a1fe38d2cd8b7b69a59d9d60/tslint.json

angular-specific rules were taken out

Blackbaud-ColbyWhite committed:
doc: include pr link in changelog

blackbaud-johnly commented on src/service/config-store.ts:
```diff
@@ -0,0 +1,56 @@
+  ['fenxt(fenxt.cdev)']: feConfig,
+  ['fenxt(fenxt.ate)']: feConfig,
+  // This is necessary as bbcrm is using the old help format, and bulkimport is a modern SPA.
+  fndlite: fndliteConfig,
```
The comment makes me assume this has something to do with BBCRM, but this pulls up the TCS help file. Is that the correct behavior?

Blackbaud-ColbyWhite replied:
this comment came from the bb-help SPA. looking through that project's history, it looks like a [commit reordered the values](https://dev.azure.com/blackbaud/Products/_git/skyux-spa-bb-help/commit/062f3a7c41275343e94c3fad3fc4da84b6f77c09?refName=refs%2Fheads%2Fmaster&path=%2Fsrc%2Fapp%2Fcomponents%2Fshared%2Fconfig-store.service.ts&_a=contents) but did not move the comment along with it. it seems like that comment is supposed to be on the `bbcrm` property based on an [earlier commit](https://dev.azure.com/blackbaud/Products/_git/skyux-spa-bb-help?_a=blame&path=%2Fsrc%2Fapp%2Fcomponents%2Fshared%2Fconfig-store.service.ts&version=GCc9a0f480111a239a9a6090cd7137418230fcf425). which makes more sense. i'll move the comment to avoid future confusion.

blackbaud-johnly commented on src/service/config-store.ts:
```diff
@@ -0,0 +1,56 @@
+import { STEWARDSHIP_MANAGEMENT_CONFIG } from './configs/stewardship';
+import { config as tcsConfig } from './configs/tcs';
+
+export type VALID_CONFIG_ID = 'admin'
```
Probably doesn't matter here, but should this list be in alphabetical order when we create something like this for a supportal? ... Also, I wonder if this would be more intuitive for the supportal if this list was a dropdown too. We could have 2 dropdowns, with this list of svcids first, then the dropdown for locales, and finally a load button. That might be more intuitive in the supportal experience where writers are probably focused on just one svcid (and its locales in the rare cases when we have locales). What do you think?

Blackbaud-ColbyWhite replied:
on the order of _this particular list_, the order doesn't _technically_ matter. this is j/ a type to help the typescript compiler understand what a valid id is. it's not generating the list in this project's demo page. but your point is still valid. keeping lists/types like this in alphabetical order makes it easier to read. i'll reorder this type.

on the order of _the list in `template.html`_ that generates the dropdown, your point is extra valid. i'll put those in a sane order. and we should do the same on the future supportal.

on the two-dropdown approach to the future supportal, agreed. that's likely a better experience. i'll go ahead and use that strategy on this demo page so you can see how that would roughly look

blackbaud-johnly commented on template.html:
```diff
@@ -1,21 +1,75 @@
+  <div class="sky-form-group">
+    <label for="locales" class="sky-control-label">Locale</label>
+    <select class="sky-form-control" id="locales">
+      <option value="">UNSPECIFIED</option>
```
```suggestion
      <option value="">Unspecified</option>
```

blackbaud-johnly commented on src/service/configs/fndlite.ts:
```diff
@@ -0,0 +1,9 @@
+export const config: HelpConfig = {
+  productId: 'fndlite',
+  hideHelpChat: true,
+  defaultHelpKey: 'tcs.html',
```
Is this correct? I don't know what the fndlite svsid is or, but should it really point to the TCS help file?

Blackbaud-ColbyWhite replied:
valid question. this is the same config from the [bb-help SPA](https://dev.azure.com/blackbaud/Products/_git/skyux-spa-bb-help?path=%2Fsrc%2Fhelp_configs%2Ffndlite.ts&version=GBmaster&_a=blame) so nothing's changing tho.

i _think_ at one point `fndlite` meant tcs, which is probably what was driving this. but i _think_ `fndlite` is _no longer_ a thing. i actually don't know.

blackbaud-johnly replied:
Makes sense. Sounds to me like we should just leave this as is. Thanks!

blackbaud-johnly commented on CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+
+- Replaced communication with bb-help SPA with functionality to open topics in a new tab. [#75](https://github.com/blackbaud/help-client/pull/75)
+- Added unload functionality. [#75](https://github.com/blackbaud/help-client/pull/75)
+- Deprecate config properties not related to topic URLs or styling. [#75](https://github.com/blackbaud/help-client/pull/75)
```
```suggestion
- Deprecated config properties that are not related to topic URLs or styling. [#75](https://github.com/blackbaud/help-client/pull/75)
```

blackbaud-johnly commented on CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+# 3.0.1-rc.0
+
+- Replaced communication with bb-help SPA with functionality to open topics in a new tab. [#75](https://github.com/blackbaud/help-client/pull/75)
```
```suggestion
- Replaced communication with the bb-help SPA to open topics in a new tab. [#75](https://github.com/blackbaud/help-client/pull/75)
```

blackbaud-johnly commented on src/service/config-store.ts:
```diff
@@ -0,0 +1,56 @@
+  chrch: churchConfig,
+  edu: eduConfig,
+  faith: faithConfig,
+  fenxt: feConfig,
```
FE NXT has a couple locale folders, but they aren't working. It only has `en-au` and `en-ca`, but I can't get either those to open through this PR.

Blackbaud-ColbyWhite replied:
interesting. it looks like i somehow copy-pasted wrong the FE config from [bb-help](https://dev.azure.com/blackbaud/Products/_git/skyux-spa-bb-help?path=%2Fsrc%2Fhelp_configs%2Ffe.ts&version=GBmaster&_a=blame) and left out the `customLocales`. will fix.

blackbaud-johnly reviewed (approved):
I commented with a few minor suggestions and some questions, but nothing too important.

Blackbaud-ColbyWhite committed:
refactor: place bbcrm comment in correct place

Blackbaud-ColbyWhite committed:
refactor: follow BB casing in demo page

Co-authored-by: John Lyons <John.Lyons@blackbaud.com>

Blackbaud-ColbyWhite committed:
doc: better grammar in changelog

Co-authored-by: John Lyons <John.Lyons@blackbaud.com>

Blackbaud-ColbyWhite committed:
doc: better grammar in changelog

Co-authored-by: John Lyons <John.Lyons@blackbaud.com>

Blackbaud-ColbyWhite committed:
refactor: order svcids

Blackbaud-ColbyWhite committed:
fix: bring back FE locales

Blackbaud-ColbyWhite committed:
deprecate: fndlite isn't a thing anymore

Blackbaud-ColbyWhite committed:
feat: use dropdown for svcids in demo

Blackbaud-ColbyWhite committed:
fix: remove container during unload

if this is not done, the subsequent load will create a _new_ container
in the DOM, resulting in duplicate containers cluttering the DOM.

Blackbaud-ColbyWhite committed:
npm version prerelease --preid=rc
